File: repertoire/concepts/data-integration/providers/enricher_plugin/video_summary.ts
// Clef Data Integration Kit - Video summarization enricher provider
// Extracts keyframes at intervals, generates transcript via whisper, combines for chapter markers.

import { execFile } from 'child_process';
import { promisify } from 'util';
import { writeFileSync, readFileSync, unlinkSync, mkdtempSync, readdirSync, rmdirSync } from 'fs';
import { tmpdir } from 'os';
import { join } from 'path';

const execFileAsync = promisify(execFile);

export const PROVIDER_ID = 'video_summary';
export const PLUGIN_TYPE = 'enricher_plugin';

export interface ContentItem {
  id: string;
  content: string;
  contentType: string;
  metadata?: Record<string, unknown>;
}

export interface EnricherConfig {
  model?: string;
  apiKey?: string;
  threshold?: number;
  options?: Record<string, unknown>;
}

export interface EnrichmentResult {
  fields: Record<string, unknown>;
  confidence: number;
  metadata?: Record<string, unknown>;
}

export interface SchemaRef {
  name: string;
  fields?: string[];
}

export interface CostEstimate {
  tokens?: number;
  apiCalls?: number;
  durationMs?: number;
}

interface Chapter {
  title: string;
  start: number;
  end: number;
}

interface Keyframe {
  timestamp: number;
  description: string;
}

interface TranscriptSegment {
  text: string;
  start: number;
  end: number;
}

async function getVideoDuration(videoPath: string): Promise<number> {
  const { stdout } = await execFileAsync('ffprobe', [
    '-v', 'quiet',
    '-print_format', 'json',
    '-show_format',
    videoPath,
  ]);
  const info = JSON.parse(stdout);
  return parseFloat(info.format?.duration ?? '0');
}

async function extractKeyframes(
  videoPath: string,
  intervalSec: number,
  outputDir: string,
): Promise<Keyframe[]> {
  // Extract frames at regular intervals using ffmpeg scene detection + interval
  await execFileAsync('ffmpeg', [
    '-i', videoPath,
    '-vf', `fps=1/${intervalSec},select='gt(scene\\,0.3)+not(mod(n\\,1))'`,
    '-vsync', 'vfr',
    '-q:v', '2',
    join(outputDir, 'keyframe_%04d.jpg'),
  ], { timeout: 120000 });

  const files = readdirSync(outputDir)
    .filter((f) => f.startsWith('keyframe_'))
    .sort();

  return files.map((file, idx) => ({
    timestamp: idx * intervalSec,
    description: `Keyframe at ${formatTimestamp(idx * intervalSec)}`,
  }));
}

async function extractAudioTrack(videoPath: string, outputPath: string): Promise<void> {
  await execFileAsync('ffmpeg', [
    '-i', videoPath,
    '-vn',
    '-acodec', 'pcm_s16le',
    '-ar', '16000',
    '-ac', '1',
    outputPath,
  ], { timeout: 120000 });
}

async function transcribeAudio(audioPath: string, language: string): Promise<TranscriptSegment[]> {
  const { stdout } = await execFileAsync('whisper', [
    audioPath,
    '--model', 'base',
    '--language', language,
    '--output_format', 'json',
    '--output_dir', tmpdir(),
  ], { timeout: 300000 });

  const baseName = audioPath.replace(/\.[^.]+$/, '');
  const jsonPath = `${baseName}.json`;

  try {
    const data = JSON.parse(readFileSync(jsonPath, 'utf-8'));
    return (data.segments ?? []).map((seg: any) => ({
      text: (seg.text ?? '').trim(),
      start: seg.start ?? 0,
      end: seg.end ?? 0,
    }));
  } catch {
    return [];
  } finally {
    try { unlinkSync(jsonPath); } catch { /* best-effort */ }
  }
}

function generateChapters(
  segments: TranscriptSegment[],
  chapterInterval: number,
  totalDuration: number,
): Chapter[] {
  const chapters: Chapter[] = [];
  let chapterStart = 0;

  while (chapterStart < totalDuration) {
    const chapterEnd = Math.min(chapterStart + chapterInterval, totalDuration);

    // Collect transcript segments within this chapter window
    const chapterSegments = segments.filter(
      (s) => s.start >= chapterStart && s.start < chapterEnd,
    );

    // Generate chapter title from first meaningful sentence or key phrases
    let title = `Chapter at ${formatTimestamp(chapterStart)}`;
    if (chapterSegments.length > 0) {
      const combinedText = chapterSegments.map((s) => s.text).join(' ');
      // Extract most significant phrase (first sentence or first N words)
      const firstSentence = combinedText.split(/[.!?]/).filter(Boolean)[0] ?? '';
      title = firstSentence.trim().slice(0, 80) || title;
    }

    chapters.push({ title, start: chapterStart, end: chapterEnd });
    chapterStart = chapterEnd;
  }

  return chapters;
}

function generateSummary(
  segments: TranscriptSegment[],
  lengths: { short: number; medium: number; long: number },
): { short: string; medium: string; long: string } {
  const fullText = segments.map((s) => s.text).join(' ');
  const sentences = fullText.split(/[.!?]+/).filter((s) => s.trim().length > 10);

  // Score sentences by position and keyword density
  const scoredSentences = sentences.map((sentence, idx) => {
    let score = 0;
    // Position bonus: first and last sentences score higher
    if (idx === 0) score += 3;
    if (idx === sentences.length - 1) score += 2;
    if (idx < sentences.length * 0.2) score += 1;

    // Length bonus: prefer medium-length sentences
    const wordCount = sentence.split(/\s+/).length;
    if (wordCount >= 8 && wordCount <= 25) score += 2;

    return { sentence: sentence.trim(), score };
  });

  scoredSentences.sort((a, b) => b.score - a.score);

  const pickSentences = (count: number): string =>
    scoredSentences
      .slice(0, count)
      .map((s) => s.sentence)
      .join('. ') + '.';

  return {
    short: pickSentences(lengths.short),
    medium: pickSentences(lengths.medium),
    long: pickSentences(lengths.long),
  };
}

function formatTimestamp(seconds: number): string {
  const h = Math.floor(seconds / 3600);
  const m = Math.floor((seconds % 3600) / 60);
  const s = Math.floor(seconds % 60);
  return h > 0
    ? `${h}:${String(m).padStart(2, '0')}:${String(s).padStart(2, '0')}`
    : `${m}:${String(s).padStart(2, '0')}`;
}

export class VideoSummaryEnricherProvider {
  async enrich(item: ContentItem, config: EnricherConfig): Promise<EnrichmentResult> {
    const chapterInterval = (config.options?.chapterInterval as number) ?? 300; // 5 minutes
    const summaryLengths = (config.options?.summaryLengths as { short: number; medium: number; long: number }) ??
      { short: 2, medium: 5, long: 10 };
    const language = (config.options?.language as string) ?? 'en';

    // Write video to temp file
    const videoPath = join(tmpdir(), `clef_video_${item.id}_${Date.now()}.mp4`);
    const audioPath = join(tmpdir(), `clef_audio_${item.id}_${Date.now()}.wav`);
    const keyframeDir = mkdtempSync(join(tmpdir(), 'clef_keyframes_'));

    writeFileSync(videoPath, Buffer.from(item.content, 'base64'));

    try {
      // Step 1: Get video duration
      const duration = await getVideoDuration(videoPath);

      // Step 2: Extract keyframes at chapter intervals
      const keyframeInterval = Math.max(10, chapterInterval / 6);
      const keyframes = await extractKeyframes(videoPath, keyframeInterval, keyframeDir);

      // Step 3: Extract audio and transcribe
      await extractAudioTrack(videoPath, audioPath);
      const segments = await transcribeAudio(audioPath, language);

      // Step 4: Generate chapters from transcript
      const chapters = generateChapters(segments, chapterInterval, duration);

      // Step 5: Generate multi-length summaries
      const summary = generateSummary(segments, summaryLengths);

      const fullTranscript = segments.map((s) => s.text).join(' ');

      return {
        fields: {
          summary,
          chapters,
          keyframes,
          transcript_text: fullTranscript,
          duration_seconds: duration,
        },
        confidence: segments.length > 0 ? 0.8 : 0.4,
        metadata: {
          provider: PROVIDER_ID,
          chapterCount: chapters.length,
          keyframeCount: keyframes.length,
          segmentCount: segments.length,
          language,
        },
      };
    } finally {
      try { unlinkSync(videoPath); } catch { /* cleanup */ }
      try { unlinkSync(audioPath); } catch { /* cleanup */ }
      try {
        readdirSync(keyframeDir).forEach((f) => unlinkSync(join(keyframeDir, f)));
        rmdirSync(keyframeDir);
      } catch { /* cleanup */ }
    }
  }

  appliesTo(schema: SchemaRef): boolean {
    const videoSchemas = ['video', 'movie', 'clip', 'recording', 'lecture', 'presentation'];
    return videoSchemas.some((s) => schema.name.toLowerCase().includes(s));
  }

  costEstimate(item: ContentItem): CostEstimate {
    const sizeKb = Buffer.byteLength(item.content, 'base64') / 1024;
    // Video: ~500KB/sec compressed. Estimate total processing time
    const estimatedDurationSec = sizeKb / 500;
    const keyframeExtractionMs = estimatedDurationSec * 100;
    const transcriptionMs = estimatedDurationSec * 33;
    const totalMs = keyframeExtractionMs + transcriptionMs + 5000;

    return {
      durationMs: Math.round(totalMs),
      apiCalls: 1,
    };
  }
}

export default VideoSummaryEnricherProvider;

File: repertoire/concepts/data-integration/providers/enricher_plugin/VideoSummary.swift
// Clef Data Integration Kit - Video summarization enricher provider
// Extracts keyframes at intervals, generates transcript via whisper, combines for chapter markers.

import Foundation

public let VideoSummaryProviderID = "video_summary"
public let VideoSummaryPluginType = "enricher_plugin"

public struct VideoChapter {
    public let title: String
    public let start: Double
    public let end: Double
}

public struct VideoKeyframe {
    public let timestamp: Double
    public let description: String
}

public final class VideoSummaryEnricherProvider {
    public init() {}

    public func enrich(item: ContentItem, config: EnricherConfig) async throws -> EnrichmentResult {
        let chapterInterval = (config.options?["chapterInterval"] as? Double) ?? 300.0
        let language = (config.options?["language"] as? String) ?? "en"
        let summaryLengths = config.options?["summaryLengths"] as? [String: Int] ?? ["short": 2, "medium": 5, "long": 10]

        guard let videoData = Data(base64Encoded: item.content) else {
            throw EnricherError.parseError("Failed to decode base64 video content")
        }

        let videoPath = NSTemporaryDirectory() + "clef_video_\(item.id)_\(ProcessInfo.processInfo.processIdentifier).mp4"
        let audioPath = NSTemporaryDirectory() + "clef_audio_\(item.id)_\(ProcessInfo.processInfo.processIdentifier).wav"
        let keyframeDir = NSTemporaryDirectory() + "clef_keyframes_\(item.id)/"

        try videoData.write(to: URL(fileURLWithPath: videoPath))
        try FileManager.default.createDirectory(atPath: keyframeDir, withIntermediateDirectories: true)

        defer {
            try? FileManager.default.removeItem(atPath: videoPath)
            try? FileManager.default.removeItem(atPath: audioPath)
            try? FileManager.default.removeItem(atPath: keyframeDir)
        }

        // Step 1: Get video duration via ffprobe
        let duration = try getVideoDuration(videoPath: videoPath)

        // Step 2: Extract keyframes using ffmpeg
        let kfInterval = max(10.0, chapterInterval / 6.0)
        let keyframes = try extractKeyframes(videoPath: videoPath, interval: kfInterval, outputDir: keyframeDir)

        // Step 3: Extract audio track and transcribe
        try extractAudioTrack(videoPath: videoPath, audioPath: audioPath)
        let segments = transcribeAudio(audioPath: audioPath, language: language)

        // Step 4: Generate chapters
        let chapters = generateChapters(segments: segments, interval: chapterInterval, totalDuration: duration)

        // Step 5: Generate summaries at multiple lengths
        let shortCount = summaryLengths["short"] ?? 2
        let mediumCount = summaryLengths["medium"] ?? 5
        let longCount = summaryLengths["long"] ?? 10
        let summary = generateSummary(segments: segments, counts: (shortCount, mediumCount, longCount))

        let fullTranscript = segments.map { $0.text }.joined(separator: " ")

        let chapterDicts: [[String: Any]] = chapters.map {
            ["title": $0.title, "start": $0.start, "end": $0.end]
        }
        let keyframeDicts: [[String: Any]] = keyframes.map {
            ["timestamp": $0.timestamp, "description": $0.description]
        }

        return EnrichmentResult(
            fields: [
                "summary": ["short": summary.short, "medium": summary.medium, "long": summary.long],
                "chapters": chapterDicts,
                "keyframes": keyframeDicts,
                "transcript_text": fullTranscript,
                "duration_seconds": duration
            ],
            confidence: segments.isEmpty ? 0.4 : 0.8,
            metadata: [
                "provider": VideoSummaryProviderID,
                "chapterCount": chapters.count,
                "keyframeCount": keyframes.count,
                "segmentCount": segments.count,
                "language": language
            ]
        )
    }

    public func appliesTo(schema: SchemaRef) -> Bool {
        let videoSchemas = ["video", "movie", "clip", "recording", "lecture", "presentation"]
        let nameLower = schema.name.lowercased()
        return videoSchemas.contains { nameLower.contains($0) }
    }

    public func costEstimate(item: ContentItem) -> CostEstimate {
        let sizeKb = Double(item.content.count) * 3.0 / 4.0 / 1024.0
        let estimatedDurationSec = sizeKb / 500.0
        let keyframeMs = estimatedDurationSec * 100.0
        let transcriptionMs = estimatedDurationSec * 33.0
        let totalMs = keyframeMs + transcriptionMs + 5000.0
        return CostEstimate(tokens: nil, apiCalls: 1, durationMs: Int(totalMs))
    }

    // MARK: - Video Processing

    private func getVideoDuration(videoPath: String) throws -> Double {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
        process.arguments = ["ffprobe", "-v", "quiet", "-print_format", "json", "-show_format", videoPath]
        let pipe = Pipe()
        process.standardOutput = pipe
        try process.run()
        process.waitUntilExit()

        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let format = json["format"] as? [String: Any],
              let durationStr = format["duration"] as? String,
              let duration = Double(durationStr) else {
            throw EnricherError.parseError("Could not read video duration")
        }
        return duration
    }

    private func extractKeyframes(videoPath: String, interval: Double, outputDir: String) throws -> [VideoKeyframe] {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
        let vfFilter = "fps=1/\(Int(interval)),select='gt(scene,0.3)+not(mod(n,1))'"
        process.arguments = [
            "ffmpeg", "-i", videoPath, "-vf", vfFilter,
            "-vsync", "vfr", "-q:v", "2", "\(outputDir)/keyframe_%04d.jpg"
        ]
        try process.run()
        process.waitUntilExit()

        let files = (try? FileManager.default.contentsOfDirectory(atPath: outputDir)) ?? []
        let sorted = files.filter { $0.hasPrefix("keyframe_") }.sorted()

        return sorted.enumerated().map { (idx, _) in
            let ts = Double(idx) * interval
            return VideoKeyframe(timestamp: ts, description: "Keyframe at \(formatTimestamp(ts))")
        }
    }

    private func extractAudioTrack(videoPath: String, audioPath: String) throws {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
        process.arguments = ["ffmpeg", "-i", videoPath, "-vn", "-acodec", "pcm_s16le", "-ar", "16000", "-ac", "1", audioPath]
        try process.run()
        process.waitUntilExit()
    }

    private func transcribeAudio(audioPath: String, language: String) -> [(text: String, start: Double, end: Double)] {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
        process.arguments = [
            "whisper", audioPath, "--model", "base", "--language", language,
            "--output_format", "json", "--output_dir", NSTemporaryDirectory()
        ]
        do { try process.run() } catch { return [] }
        process.waitUntilExit()

        let stem = (audioPath as NSString).deletingPathExtension
        let jsonPath = NSTemporaryDirectory() + ((stem as NSString).lastPathComponent) + ".json"
        defer { try? FileManager.default.removeItem(atPath: jsonPath) }

        guard let data = try? Data(contentsOf: URL(fileURLWithPath: jsonPath)),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let segs = json["segments"] as? [[String: Any]] else { return [] }

        return segs.compactMap { seg in
            guard let text = seg["text"] as? String else { return nil }
            let start = (seg["start"] as? Double) ?? 0.0
            let end = (seg["end"] as? Double) ?? 0.0
            return (text.trimmingCharacters(in: .whitespaces), start, end)
        }
    }

    // MARK: - Chapter & Summary Generation

    private func generateChapters(
        segments: [(text: String, start: Double, end: Double)],
        interval: Double, totalDuration: Double
    ) -> [VideoChapter] {
        var chapters: [VideoChapter] = []
        var chapterStart = 0.0

        while chapterStart < totalDuration {
            let chapterEnd = min(chapterStart + interval, totalDuration)
            let chapterSegs = segments.filter { $0.start >= chapterStart && $0.start < chapterEnd }

            let title: String
            if !chapterSegs.isEmpty {
                let combined = chapterSegs.map { $0.text }.joined(separator: " ")
                let firstSentence = combined.components(separatedBy: CharacterSet(charactersIn: ".!?"))
                    .first(where: { !$0.trimmingCharacters(in: .whitespaces).isEmpty }) ?? ""
                title = String(firstSentence.trimmingCharacters(in: .whitespaces).prefix(80))
            } else {
                title = "Chapter at \(formatTimestamp(chapterStart))"
            }

            chapters.append(VideoChapter(title: title, start: chapterStart, end: chapterEnd))
            chapterStart = chapterEnd
        }
        return chapters
    }

    private func generateSummary(
        segments: [(text: String, start: Double, end: Double)],
        counts: (Int, Int, Int)
    ) -> (short: String, medium: String, long: String) {
        let fullText = segments.map { $0.text }.joined(separator: " ")
        let sentences = fullText.components(separatedBy: CharacterSet(charactersIn: ".!?"))
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { $0.split(separator: " ").count >= 3 }

        var scored: [(sentence: String, score: Double)] = sentences.enumerated().map { (idx, sent) in
            var score = 0.0
            if idx == 0 { score += 3.0 }
            if idx == sentences.count - 1 { score += 2.0 }
            if Double(idx) < Double(sentences.count) * 0.2 { score += 1.0 }
            let wc = sent.split(separator: " ").count
            if wc >= 8 && wc <= 25 { score += 2.0 }
            return (sent, score)
        }
        scored.sort { $0.score > $1.score }

        func pick(_ n: Int) -> String {
            scored.prefix(n).map { $0.sentence }.joined(separator: ". ") + "."
        }

        return (pick(counts.0), pick(counts.1), pick(counts.2))
    }

    private func formatTimestamp(_ seconds: Double) -> String {
        let total = Int(seconds)
        let h = total / 3600
        let m = (total % 3600) / 60
        let s = total % 60
        if h > 0 {
            return String(format: "%d:%02d:%02d", h, m, s)
        }
        return String(format: "%d:%02d", m, s)
    }
}
